store/store.go

```diff
@@ -400,12 +392,61 @@ func (d MongoStoreClient) GetDeviceData(p *Params) StorageIterator {
 	// closes session when the iterator is closed. See ClosingSessionIterator below.
 	session := d.session.Copy()
 
-	iter := mgoDataCollection(session).
-		Find(generateMongoQuery(p)).
-		Sort(p.Sort...).
-		Limit(p.Limit).
-		Select(removeFieldsForReturn).
-		Iter()
+	var iter *mgo.Iter
+
+	if p.Latest {
+		// Create an $aggregate query to return the latest of each `type` requested
+		// that matches the query parameters
+		pipeline := []bson.M{
+			{
+				"$match": generateMongoQuery(p),
+			},
+			{
+				"$sort": bson.M{
+					"type": 1,
+					"time": -1,
+				},
+			},
+			{
+				"$group": bson.M{
+					"_id": bson.M{
+						"type": "$type",
+					},
+					"groupId": bson.M{
+						"$first": "$_id",
+					},
+				},
+			},
+			{
+				"$lookup": bson.M{
+					"from":         "deviceData",
+					"localField":   "groupId",
+					"foreignField": "_id",
+					"as":           "latest_doc",
+				},
```

Comment:
Left join.  Nice.  Is there an index on the join field?

Reply:
There sure is! `_id` is the default primary key in MongoDB